src/TextInput.js

```diff
@@ -22,27 +25,64 @@ const sizeVariants = variant({
   }
 })
 
-const TextInput = styled.input.attrs(({theme, variant, block, ...rest}) => ({
-  type: 'text',
-  ...rest
-}))`
+const TextInput = ({icon, type, className, width, theme, ...rest}) => {
+  const wrapperClasses = classnames(className, 'TextInput-wrapper')
+  const inputClasses = classnames('TextInput-input', icon ? 'input-icon' : 'input-no-icon')
+  return (
+    <Wrapper width={width} className={wrapperClasses} theme={theme}>
+      {icon && <Octicon className="TextInput-icon" icon={icon} />}
+      <input className={inputClasses} type={type || 'text'} {...rest} />
+    </Wrapper>
+  )
+}
+
+const Wrapper = styled.span`
+  display: inline-flex;
+  align-items: stretch;
   min-height: 34px;
-  padding: 6px ${get('space.2')}px;
   font-size: ${get('fontSizes.2')}px;
   line-height: 20px;
   color: ${get('colors.gray.9')};
   vertical-align: middle;
-  background-color: ${get('colors.white')};
   background-repeat: no-repeat; // Repeat and position set for form states (success, error, etc)
   background-position: right 8px center; // For form validation. This keeps images 8px from right and centered vertically.
   border: 1px solid ${get('colors.gray.3')};
   border-radius: ${get('radii.1')}px;
   outline: none;
   box-shadow: ${get('shadows.formControl')};
 
-  &:focus {
+  .TextInput-icon {
```

Comment:
Did you consider making the `<input>` a styled component and having the styles live there? I generally try to use class names as little as possible since we can't have control over whether or not the consuming application will have conflicting/similar class names. Maybe we need to come up with a system for naming classes when needed - something like `pc-TextInput` to make sure it doesn't conflict with anything in Primer CSS?